File: backend/trendlens/rag_engine.py
"""
trendlens/rag_engine.py
RAG engine using MongoDB data and TF-IDF similarity.
No external LLM APIs.
"""

import logging
import math
import re
from typing import Any, Dict, List, Optional

logger = logging.getLogger(__name__)


def text_to_tfidf_vector(text: str) -> Dict[str, float]:
    """Convert text to a simple TF vector."""
    words = re.findall(r'[a-z]{3,}', text.lower())
    if not words:
        return {}
    tf: Dict[str, float] = {}
    for w in words:
        tf[w] = tf.get(w, 0) + 1
    total = len(words)
    for w in tf:
        tf[w] /= total
    return tf


def cosine_similarity_tfidf(vec1: Dict[str, float], vec2: Dict[str, float]) -> float:
    """Compute cosine similarity between two TF vectors."""
    if not vec1 or not vec2:
        return 0.0
    dot = sum(vec1.get(k, 0) * vec2.get(k, 0) for k in set(vec1) & set(vec2))
    norm1 = math.sqrt(sum(v ** 2 for v in vec1.values()))
    norm2 = math.sqrt(sum(v ** 2 for v in vec2.values()))
    if norm1 == 0 or norm2 == 0:
        return 0.0
    return dot / (norm1 * norm2)


class RAGEngine:
    """RAG engine that retrieves similar high-performing posts from MongoDB."""

    def find_similar_posts(
        self,
        caption: str,
        category: str = "",
        top_k: int = 5,
    ) -> List[Dict[str, Any]]:
        """Find similar high-performing posts."""
        if not caption.strip():
            return []

        try:
            return self._mongodb_search(caption, category, top_k)
        except Exception as exc:
            logger.warning("MongoDB RAG search failed: %s", exc)
            return self._fallback_search(caption, category, top_k)

    def _mongodb_search(self, caption: str, category: str, limit: int) -> List[Dict[str, Any]]:
        """Search MongoDB for similar posts."""
        from trendlens.database import (
            DatabaseManager,
            GroundTruthRepository,
            PostsRepository,
            EmbeddingsRepository,
        )

        # Check DB connection
        db_mgr = DatabaseManager()
        if not db_mgr.health_check():
            return self._fallback_search(caption, category, limit)

        # Try vector search on embeddings collection
        try:
            emb_repo = EmbeddingsRepository()
            query_embedding = self._caption_to_embedding(caption)
            results = emb_repo.vector_search(
                embedding=query_embedding,
                limit=limit,
                filter={"category": category} if category else None,
            )
            if results:
                return [
                    {
                        "caption": r.get("caption", ""),
                        "engagement_rate": r.get("engagement_rate", 0),
                        "category": r.get("category", ""),
                        "hashtags": r.get("hashtags", []),
                        "has_cta": r.get("has_cta", False),
                        "has_price": r.get("has_price", False),
                        "similarity_score": r.get("score", 0),
                        "source": "mongodb_vector",
                    }
                    for r in results
                ]
        except Exception:
            pass

        # Fallback: TF-IDF similarity
        gt_repo = GroundTruthRepository()
        query: Dict[str, Any] = {}
        if category:
            query["category"] = category

        gt_data = gt_repo.find_many(query, sort=[("engagement_rate", -1)], limit=100)
        if not gt_data:
            posts_repo = PostsRepository()
            gt_data = posts_repo.find_many(query, sort=[("engagement_rate", -1)], limit=100)

        if not gt_data:
            return self._fallback_search(caption, category, limit)

        query_vec = text_to_tfidf_vector(caption)
        scored = []
        for doc in gt_data:
            doc_caption = doc.get("caption", "")
            doc_vec = text_to_tfidf_vector(doc_caption)
            sim = cosine_similarity_tfidf(query_vec, doc_vec)
            scored.append((sim, doc))

        scored.sort(key=lambda x: x[0], reverse=True)
        results = []
        for sim, doc in scored[:limit]:
            if sim > 0.01:
                results.append({
                    "caption": doc.get("caption", ""),
                    "engagement_rate": doc.get("engagement_rate", 0),
                    "category": doc.get("category", ""),
                    "hashtags": doc.get("hashtags", []),
                    "has_cta": doc.get("has_cta", False),
                    "has_price": doc.get("has_price", False),
                    "similarity_score": round(sim, 4),
                    "source": "mongodb_tfidf",
                })
        return results

    def _fallback_search(self, caption: str, category: str, limit: int) -> List[Dict[str, Any]]:
        """Fallback heuristic examples when MongoDB is unavailable."""
        examples = {
            "cake": [
                {"caption": "Beautiful custom wedding cake! DM to order. #CakeKampala #WeddingCake UGX 150,000", "engagement_rate": 0.12, "category": "cake", "hashtags": ["CakeKampala", "WeddingCake"], "has_cta": True, "has_price": True, "similarity_score": 0.5, "source": "heuristic"},
                {"caption": "Fresh chocolate layer cake UGX 85,000. WhatsApp 0700 123456. #ChocolateCake #KampalaBakery", "engagement_rate": 0.09, "category": "cake", "hashtags": ["ChocolateCake", "KampalaBakery"], "has_cta": True, "has_price": True, "similarity_score": 0.4, "source": "heuristic"},
            ],
            "bakery": [
                {"caption": "Fresh bread every morning! Whole wheat UGX 5,000. #FreshBread #KampalaBakery", "engagement_rate": 0.08, "category": "bakery", "hashtags": ["FreshBread", "KampalaBakery"], "has_cta": True, "has_price": True, "similarity_score": 0.45, "source": "heuristic"},
            ],
            "restaurant": [
                {"caption": "Lunch special: Matooke + G-nut sauce + rice UGX 15,000! #LunchSpecial #UgandanFood", "engagement_rate": 0.11, "category": "restaurant", "hashtags": ["LunchSpecial", "UgandanFood"], "has_cta": True, "has_price": True, "similarity_score": 0.5, "source": "heuristic"},
            ],
            "general": [
                {"caption": "Support local! Buy fresh produce directly from Ugandan farmers. #BuyLocal #UgandanFarmers", "engagement_rate": 0.07, "category": "general", "hashtags": ["BuyLocal", "UgandanFarmers"], "has_cta": True, "has_price": False, "similarity_score": 0.35, "source": "heuristic"},
            ],
        }
        cat = category if category in examples else "general"
        return examples.get(cat, [])[:limit]

    def _caption_to_embedding(self, caption: str) -> List[float]:
        """Generate 384-dim embedding from caption text."""
        dim = 384
        words = re.findall(r'[a-z]{3,}', caption.lower())
        embedding = [0.0] * dim
        for word in words:
            hash_val = 0
            for ch in word:
                hash_val = ((hash_val << 5) - hash_val + ord(ch)) | 0
            idx = abs(hash_val) % dim
            embedding[idx] += 1
        embedding[dim - 1] = len(words) / 50
        embedding[dim - 2] = len(re.findall(r'#', caption)) / 15
        embedding[dim - 3] = 1.0 if re.search(r'ugx|ush|\$', caption, re.I) else 0.0
        embedding[dim - 4] = 1.0 if re.search(r'dm|whatsapp|link in bio|order', caption, re.I) else 0.0
        norm = math.sqrt(sum(v * v for v in embedding)) or 1
        return [round(v / norm, 6) for v in embedding]

File: src/lib/ai/rag-engine.ts
/**
 * TrendLens AI v6.0 — RAG Engine (Enhanced Semantic Edition)
 * Retrieval-Augmented Generation using MongoDB Atlas Vector Search.
 * 
 * Key improvements over v1:
 * - N-gram + semantic vocabulary embeddings (not just unigram hashing)
 * - Silent LLM embedding generation for richer semantic representation
 * - Graceful fallback: LLM embedding → enhanced TF → text-based search
 * - Category-aware similarity weighting
 */

import { EmbeddingsRepository, healthCheck, getCollection } from '../db/client';
import { RagInsight, VectorSearchResult, CaptionFeatures } from './types';
import { getCategoryRule } from './category-rules';

// ─── Semantic Vocabulary with Category Weighting ─────────────────────────────

const FOOD_VOCABULARY = [
  'cake', 'bread', 'pastry', 'bakery', 'restaurant', 'food', 'meal', 'dish',
  'ugx', 'delivery', 'order', 'fresh', 'homemade', 'delicious', 'special',
  'kampala', 'uganda', 'birthday', 'wedding', 'custom', 'organic', 'local',
  'breakfast', 'lunch', 'dinner', 'snack', 'dessert', 'drink', 'coffee',
  'chocolate', 'vanilla', 'chicken', 'beef', 'fish', 'rice', 'matooke',
  'whatsapp', 'dm', 'link', 'price', 'starting', 'limited', 'offer',
  'morning', 'evening', 'today', 'new', 'best', 'top', 'premium',
  'rolex', 'luwombo', 'chapati', 'mandazi', 'tilapia', 'fondant',
  'sourdough', 'croissant', 'cupcake', 'icing', 'ganache', 'bbq',
  'grilled', 'steamed', 'fried', 'baked', 'traditional', 'street food',
  'catering', 'event', 'party', 'celebration', 'introduction', 'ceremony',
];

const CATEGORY_KEYWORDS: Record<string, string[]> = {
  cake: ['cake', 'wedding', 'birthday', 'cupcake', 'fondant', 'icing', 'ganache', 'red velvet', 'buttercream', 'tier', 'custom design'],
  bakery: ['bread', 'pastry', 'bakery', 'sourdough', 'croissant', 'dough', 'flour', 'roll', 'mandazi', 'chapati', 'fresh baked'],
  restaurant: ['restaurant', 'menu', 'dish', 'meal', 'lunch', 'dinner', 'dining', 'chef', 'local food', 'traditional', 'rolex', 'luwombo', 'matooke', 'tilapia'],
  general: ['uganda', 'kampala', 'local', 'business', 'market', 'deal', 'offer', 'quality', 'support'],
};

// ─── Enhanced Embedding Generation ──────────────────────────────────────────

export function generateEnhancedEmbedding(text: string, dimensions: number = 384): number[] {
  if (!text) return new Array(dimensions).fill(0);

  const lower = text.toLowerCase();
  const words = lower.split(/\s+/).filter(w => w.length > 1);
  const embedding = new Array(dimensions).fill(0);

  // 1. Unigram features with position diversity
  for (let i = 0; i < words.length; i++) {
    const word = words[i];
    for (let j = 0; j < 3; j++) {
      const hash = multiHash(`${word}_${j}`) % (dimensions - 20);
      embedding[hash] += 1;
    }
  }

  // 2. Bigram features for richer semantic capture
  for (let i = 0; i < words.length - 1; i++) {
    const bigram = `${words[i]}_${words[i + 1]}`;
    const hash = multiHash(bigram) % (dimensions - 20);
    embedding[hash] += 0.7;
  }

  // 3. Trigram features for phrase-level semantics
  for (let i = 0; i < words.length - 2; i++) {
    const trigram = `${words[i]}_${words[i + 1]}_${words[i + 2]}`;
    const hash = multiHash(trigram) % (dimensions - 20);
    embedding[hash] += 0.4;
  }

  // 4. Semantic vocabulary mapping with weighted features
  for (let i = 0; i < FOOD_VOCABULARY.length && i < (dimensions - 20) / 3; i++) {
    if (lower.includes(FOOD_VOCABULARY[i])) {
      embedding[i] += 2.0;
    }
  }

  // 5. Category-specific keyword weighting
  let bestCategory = 'general';
  let bestScore = 0;
  for (const [cat, keywords] of Object.entries(CATEGORY_KEYWORDS)) {
    const score = keywords.filter(kw => lower.includes(kw)).length;
    if (score > bestScore) {
      bestScore = score;
      bestCategory = cat;
    }
  }
  // One-hot encode category in dedicated slots
  const categoryOffset = Math.floor((dimensions - 20) * 0.7);
  const catIndex = ['cake', 'bakery', 'restaurant', 'general'].indexOf(bestCategory);
  if (catIndex >= 0) embedding[categoryOffset + catIndex] = 1;

  // 6. Structural features in reserved last slots
  const hashCount = (text.match(/#\w+/g) || []).length;
  embedding[dimensions - 1] = Math.min(1, words.length / 60);       // Caption length
  embedding[dimensions - 2] = Math.min(1, hashCount / 15);          // Hashtag count
  embedding[dimensions - 3] = /ugx|ush|\$|price|starting/i.test(text) ? 1 : 0;  // Price
  embedding[dimensions - 4] = /dm|whatsapp|order|link in bio|call/i.test(text) ? 1 : 0;  // CTA
  embedding[dimensions - 5] = (text.match(/[\u{1F600}-\u{1F64F}]/gu) || []).length / 5;  // Emoji
  embedding[dimensions - 6] = /call|reserve|book|visit/i.test(text) ? 1 : 0;  // Secondary CTA
  embedding[dimensions - 7] = /free delivery|delivery|takeaway/i.test(text) ? 1 : 0;  // Delivery
  embedding[dimensions - 8] = /limited|flash sale|today only|while stock/i.test(text) ? 1 : 0;  // Urgency
  embedding[dimensions - 9] = /featured|voted|best|award/i.test(text) ? 1 : 0;  // Social proof
  embedding[dimensions - 10] = /ugx\s*\d/i.test(text) ? 1 : 0;  // Price with number
  embedding[dimensions - 11] = /0700|0772|0312|0780/i.test(text) ? 1 : 0;  // Ugandan phone
  embedding[dimensions - 12] = bestScore / Math.max(1, CATEGORY_KEYWORDS[bestCategory]?.length || 1);  // Category confidence
  embedding[dimensions - 13] = hashCount >= 5 && hashCount <= 10 ? 1 : 0;  // Optimal hashtag range
  embedding[dimensions - 14] = Math.min(1, (text.match(/[.!?]/g) || []).length / 5);  // Sentence count
  embedding[dimensions - 15] = words.filter(w => w === w.toUpperCase() && w.length > 2).length / 5;  // Emphasis words

  // Normalize
  const norm = Math.sqrt(embedding.reduce((sum, v) => sum + v * v, 0));
  if (norm > 0) {
    for (let i = 0; i < embedding.length; i++) {
      embedding[i] /= norm;
    }
  }

  return embedding;
}

// Backward-compatible alias
export const generateSimpleEmbedding = generateEnhancedEmbedding;

function multiHash(str: string): number {
  let hash = 0;
  for (let i = 0; i < str.length; i++) {
    const char = str.charCodeAt(i);
    hash = ((hash << 5) - hash) + char;
    hash = hash & hash;
  }
  return Math.abs(hash);
}

// ─── Silent LLM Embedding Enhancement ──────────────────────────────────────

/**
 * Uses the silent LLM to generate a semantic summary of the caption,
 * then blends it with the TF-based embedding for richer semantic representation.
 * Falls back to enhanced TF embedding if LLM is unavailable.
 */
async function generateLLMEnhancedEmbedding(
  text: string,
  category: string,
  dimensions: number = 384,
): Promise<number[]> {
  const baseEmbedding = generateEnhancedEmbedding(text, dimensions);

  try {
    // Dynamic import to avoid issues if Z-AI is not configured
    const ZAI = (await import('z-ai-web-dev-sdk')).default;
    const zai = await ZAI.create();

    const response = await zai.chat.completions.create({
      messages: [
        {
          role: 'system',
          content: 'You are a social media analytics assistant for Ugandan food businesses. Extract key semantic concepts from the caption. Return ONLY a comma-separated list of 10-15 descriptive keywords/phrases that capture the meaning, intent, and context. No explanations.',
        },
        {
          role: 'user',
          content: `Extract semantic keywords from this Ugandan food business social media post (category: ${category}):\n\n"${text}"`,
        },
      ],
      max_tokens: 150,
    });

    const content = response.choices?.[0]?.message?.content;
    if (!content) return baseEmbedding;

    // Parse LLM-generated keywords and blend into embedding
    const llmKeywords = content.toLowerCase().split(',').map(k => k.trim()).filter(k => k.length > 2);
    
    // Create a secondary embedding from LLM keywords
    const llmEmbedding = new Array(dimensions).fill(0);
    for (const keyword of llmKeywords) {
      const words = keyword.split(/\s+/);
      for (const word of words) {
        const hash = multiHash(word) % (dimensions - 20);
        llmEmbedding[hash] += 1.5;
      }
    }

    // Normalize LLM embedding
    const llmNorm = Math.sqrt(llmEmbedding.reduce((s, v) => s + v * v, 0)) || 1;
    for (let i = 0; i < llmEmbedding.length; i++) {
      llmEmbedding[i] /= llmNorm;
    }

    // Blend: 70% TF + 30% LLM (LLM adds semantic richness without dominating)
    const blended = baseEmbedding.map((v, i) => v * 0.7 + llmEmbedding[i] * 0.3);

    // Re-normalize
    const blendNorm = Math.sqrt(blended.reduce((s, v) => s + v * v, 0)) || 1;
    return blended.map(v => Number((v / blendNorm).toFixed(6)));

  } catch {
    // LLM unavailable — use enhanced TF embedding
    return baseEmbedding;
  }
}

// ─── RAG Search ────────────────────────────────────────────────────────────

export async function searchSimilarPosts(
  caption: string,
  category: string,
  limit: number = 5,
): Promise<VectorSearchResult[]> {
  try {
    // Try LLM-enhanced embedding first, fall back to enhanced TF
    const embedding = await generateLLMEnhancedEmbedding(caption, category);
    const repo = new EmbeddingsRepository();

    try {
      const results = await repo.vectorSearch(embedding, limit, { category });
      return results as VectorSearchResult[];
    } catch {
      // Vector search might not be available (no index created yet)
      // Fallback to enhanced text-based search
      return await enhancedTextSearch(caption, category, limit);
    }
  } catch {
    return [];
  }
}

/**
 * Enhanced text-based search using MongoDB text search + keyword matching.
 * This is the fallback when Atlas Vector Search isn't configured.
 */
async function enhancedTextSearch(
  caption: string,
  category: string,
  limit: number,
): Promise<VectorSearchResult[]> {
  try {
    const dbConnected = await healthCheck();
    if (!dbConnected) return [];

    const collection = await getCollection('embeddings');
    
    // Extract meaningful keywords from caption
    const keywords = caption.toLowerCase()
      .replace(/[^a-z\s]/g, '')
      .split(/\s+/)
      .filter(w => w.length > 3 && !['that', 'this', 'with', 'from', 'have', 'been', 'they', 'will', 'what', 'when', 'make', 'like', 'just', 'over', 'such', 'take', 'than', 'them', 'very', 'also', 'into', 'more', 'some', 'could', 'time', 'these', 'about', 'which', 'their', 'would', 'there', 'other', 'after', 'most', 'being', 'where'].includes(w))
      .slice(0, 8);

    if (keywords.length === 0) return [];

    // Use MongoDB $or with regex for flexible matching
    const results = await collection.find({
      category,
      $or: keywords.map(kw => ({ caption: { $regex: kw, $options: 'i' } })),
    }).sort({ engagement_rate: -1 }).limit(limit).toArray();

    return results.map(r => ({
      _id: r._id?.toString() || '',
      caption: (r.caption as string) || '',
      engagementRate: (r.engagement_rate as number) || 0,
      category: (r.category as string) || category,
      score: 0.5,
      hashtags: (r.hashtags as string[]) || [],
      hasCta: (r.has_cta as boolean) || false,
      hasPrice: (r.has_price as boolean) || false,
    }));
  } catch {
    return [];
  }
}

// ─── RAG Insight Generation ────────────────────────────────────────────────

export function generateRagInsights(
  similarPosts: VectorSearchResult[],
  captionFeatures: CaptionFeatures,
  category: string,
): RagInsight[] {
  if (similarPosts.length === 0) return [];

  return similarPosts.slice(0, 5).map(post => {
    const patterns: string[] = [];
    const rules = getCategoryRule(category);

    if (post.hasCta && !captionFeatures.hasCta) {
      patterns.push('Uses a call-to-action');
    }
    if (post.hasPrice && !captionFeatures.hasPrice) {
      patterns.push('Includes pricing info');
    }
    if (post.hashtags.length >= rules.idealHashtags && captionFeatures.hashtagCount < rules.idealHashtags) {
      patterns.push(`Uses ${post.hashtags.length}+ hashtags`);
    }
    if (post.engagementRate > 0.7) {
      patterns.push('High engagement rate');
    }

    let takeaway = '';
    if (patterns.length > 0) {
      takeaway = `This similar post ${patterns.slice(0, 2).join(' and ')}, achieving ${Math.round(post.engagementRate * 100)}% engagement.`;
    } else {
      takeaway = `This ${category} post achieved ${Math.round(post.engagementRate * 100)}% engagement.`;
    }

    return {
      postId: post._id,
      caption: post.caption.slice(0, 200) + (post.caption.length > 200 ? '...' : ''),
      engagementRate: post.engagementRate,
      category: post.category,
      similarity: post.score,
      keyPatterns: patterns,
      takeaway,
    };
  });
}

// ─── Store Embedding ───────────────────────────────────────────────────────

export async function storePostEmbedding(
  postId: string,
  caption: string,
  category: string,
  engagementRate: number,
  hashtags: string[] = [],
  hasCta: boolean = false,
  hasPrice: boolean = false,
): Promise<void> {
  try {
    // Try LLM-enhanced embedding, fall back to enhanced TF
    const embedding = await generateLLMEnhancedEmbedding(caption, category);
    const repo = new EmbeddingsRepository();
    await repo.storeEmbedding({
      post_id: postId,
      caption,
      category,
      engagement_rate: engagementRate,
      embedding,
      hashtags,
      has_cta: hasCta,
      has_price: hasPrice,
    });
  } catch {
    // Non-critical — don't fail the evaluation
  }
}
